Improve data fetching in Order Details, to avoid I/O performance on the main thread

pmusolino opened the pull request:
Closes: #14923


## Description
We observed that while opening product details, it's easy to see the following warning, generated by multiple reads on storage (Core Data). Also, this issue sometimes generate a lag opening the Order Detail screen.

![Image](https://github.com/user-attachments/assets/a81c4d76-5d67-4398-b8c1-3c1cbcf501a5)

This warning is occurring because we're making [a lot of different fetch requests](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/ViewModels/Order%20Details/OrderDetailsResultsControllers.swift) when loading the details, causing the hang issue.

I solved the issue by removing two fetches from `OrderDetailsResultsControllers`:
- updating the `feeLines` property to return the relationship from `order.fees` instead, of using a Results Controller to fetch the order fees.
- I added the `shippingLabels` property to `Order` which was connected in Storage with a relationship, but not present inside the Networking model.
     - Update `Networking.generated.swift`, `Models+Copiable.generated.swift`, and `Order.swift` to include `shippingLabels` property in `Order` struct.
     - Update `OrderDetailsResultsControllers.swift` by removing `shippingLabelResultsController` and using `order.shippingLabels` instead.
     - Update various files to incorporate `shippingLabels` property in Order creation and manipulation.
     - Add `Sendable` conformance to `ShippingLabel`, `ShippingLabelAddress`, `ShippingLabelRefund`, `ShippingLabelRefundStatus`, and `ShippingLabelStatus` structs and enums for concurrency safety.
     - Update tests in `OrderDetailsDataSourceTests.swift` to handle `shippingLabels` within `Order` instances.

Those were the only relationships with Order. In my opinion, we could potentially improve performance further by handling some fetches externally (e.g., fetching all order statuses). However, the current implementation is structured such that each view/model doesn't have visibility beyond the order itself. While this approach is possible, we need to discuss how deep we want to go with such changes. The current solution, however, does address the I/O problem effectively.

## Steps to reproduce
Open multiple times different Order Details. You can notice a purple warning mentioning `Performing 1/0 on the main thread by reading or writing to a database can cause slow launches.` in `ResultsController`.

## Testing information
You should not be able anymore to replicate the warning after these changes.

- Make sure you are able to see “Custom amounts” section adding a fee, and that editing the fee update the fee title/amount.
- Test that Shipping Labels in Order Details works like before if the plugin is installed: fetching/creation/update.
      - Ensure the "Create Shipping Label" button is visible when the order is eligible for shipping label creation and has no existing labels.
      - Ensure the "Create Shipping Label" button is not visible when the order is eligible but already has shipping labels.
      - Ensure the "More" button is visible in the product section when the order is eligible and has no refunded shipping labels.
      - Ensure the "More" button is visible in the product section when the order is eligible and has refunded shipping labels.
      - Ensure the "More" button is not visible in the product section when the order is not eligible for shipping label creation.
      - Ensure that refunded shipping labels are correctly identified and handled, and that the UI reflects this state appropriately.
      - Confirm that the order is correctly marked as eligible for shipping label creation based on predefined criteria (e.g., order status, payment status).
      - Ensure that any changes to the shipping labels (e.g., addition, refund) are immediately reflected in the UI without requiring a manual refresh.






---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

pmusolino committed:
update: `feeLines` property to return `order.fees` instead, while removed `feeLinesResultsController` and related logic

pmusolino committed:
Add `shippingLabels` property to `Order` which was connected in Storage, but not present inside the Networking model. Doing this to reduce number of `ResultsController` from `OrderDetailsResultsControllers`:
- Update `Networking.generated.swift`, `Models+Copiable.generated.swift`, and `Order.swift` to include `shippingLabels` property in `Order` struct
- Update `OrderDetailsResultsControllers.swift` by removing `shippingLabelResultsController` and using `order.shippingLabels` instead
- Update various files to incorporate `shippingLabels` property in Order creation and manipulation
- Add `Sendable` conformance to `ShippingLabel`, `ShippingLabelAddress`, `ShippingLabelRefund`, `ShippingLabelRefundStatus`, and `ShippingLabelStatus` structs and enums for concurrency safety
- Update tests in `OrderDetailsDataSourceTests.swift` to handle `shippingLabels` within `Order` instances

pmusolino committed:
Merge commit 'f12dbd8aeff195fe6da618d23550f9d8d6d24b7f' into feat/batch-fetching-in-order-details

pmusolino committed:
update: relese-notes

pmusolino committed:
update: add PR link to release-notes

selanthiraiyan commented:
Thanks for working on this, @pmusolino! 

Not directly related to this PR: While attempting to test this PR I noticed that the "Add Extension To Store" from the following screen doesn't work. I also noticed that the code is not linked to any action. [Code link](https://github.com/woocommerce/woocommerce-ios/blob/93628d3bc3d81612f11ec602045942a11c8d9298/WooCommerce/Classes/ViewRelated/Orders/Order%20Details/WCShip%20Installation%20Process/WCShipCTAView.swift#L59)


<img src="https://github.com/user-attachments/assets/69c7afd5-6d11-4f12-8062-90890671f694" width="320"/>


Video recording:


https://github.com/user-attachments/assets/0be2da2d-7715-4416-8571-4bc133b172d5

selanthiraiyan commented:
I noticed one more issue that is probably not related to this PR.

I have installed WooCommerce Shipping on my JN store, but I still see the "Get WooCommerce Shipping" banner in the Order Details screen. 

https://github.com/user-attachments/assets/19d2cdbd-ae2b-4ab3-905b-facfc5e581c5


I have tried restarting the app (stop and run again from Xcode), but the banner doesn't go away.

selanthiraiyan reviewed (commented):
The following issue happens in trunk as well and I don't think this is related to this PR. IMO, this looks related to this CUW post - peaMlT-Yo-p2

After creating a shipping label on my JN store I started facing issues with loading orders.

I tried emptying the orders creating Woo Smooth Generator from my JN store and kept only the order that I manually created from the mobile app. But still the mobile app fails to load that order and shows the following decoding error in logs. 

```
⛔️ Error synchronizing orders: typeMismatch(Swift.Int64, Swift.DecodingError.Context(codingPath: [CodingKeys(stringValue: "data", intValue: nil), _JSONKey(stringValue: "Index 4", intValue: 4), CodingKeys(stringValue: "line_items", intValue: nil), _JSONKey(stringValue: "Index 0", intValue: 0), CodingKeys(stringValue: "taxes", intValue: nil), _JSONKey(stringValue: "Index 0", intValue: 0), CodingKeys(stringValue: "id", intValue: nil)], debugDescription: "Expected to decode Int64 but found a string instead.", underlyingError: nil))
```

My JN site has the following plugins installed. 
<img width="1531" alt="Screenshot 2025-01-29 at 2 26 09 PM" src="https://github.com/user-attachments/assets/eeca8890-3a14-4087-87bf-efb54d27c71a" />


Please let me know if you need an invitation to my JN site.

rachelmcr commented:
> Not directly related to this PR: While attempting to test this PR I noticed that the "Add Extension To Store" from the following screen doesn't work. I also noticed that the code is not linked to any action.

FWIW the "Get WooCommerce Shipping" banner with this install prompt is behind the `shippingLabelsOnboardingM1` feature flag. I believe that was an older Kiwi project that never shipped; to avoid confusion it's probably worth removing the code or entirely disabling that feature flag.

selanthiraiyan commented:
@pmusolino  While trying to Refund a shipping label, I found that the "Request refund" button doesn't do anything. 

![Simulator Screen Recording - iPhone 15 Pro Max - 2025-01-30 at 12 11 03](https://github.com/user-attachments/assets/67ec4b85-4638-4e5a-b025-1cd8d1ee63ee)

selanthiraiyan reviewed (commented):
@pmusolino 👋 

Quoting the following from the "Testing Information" in the PR description

> - Ensure that any changes to the shipping labels (e.g., addition, refund) are immediately reflected in the UI without requiring a manual refresh.


I noticed that the order detail page doesn't reflect the newly purchased shipping label unless I pull down to refresh. Even after I pull to refresh, I see the "Create Shipping Label" button, and it goes away after another pull down to refresh. 

https://github.com/user-attachments/assets/1c50ec9b-49c3-40e6-9cbd-d92c8ebef088

pmusolino commented:
> I noticed one more issue that is probably not related to this PR.
> I have installed WooCommerce Shipping on my JN store, but I still see the "Get WooCommerce Shipping" banner in the Order Details screen.

I filled an [issue](https://github.com/woocommerce/woocommerce-ios/issues/15019) here about removing the feature flag.

pmusolino committed:
Merge branch 'trunk' into feat/batch-fetching-in-order-details

pmusolino commented:
> @pmusolino While trying to Refund a shipping label, I found that the "Request refund" button doesn't do anything.

This doesn't work in `trunk` either because I think it's still a WIP under a feature flag. cc @rachelmcr

pmusolino commented:
> I noticed that the order detail page doesn't reflect the newly purchased shipping label unless I pull down to refresh. Even after I pull to refresh, I see the "Create Shipping Label" button, and it goes away after another pull down to refresh.

I tested it by creating the shipping label from the web, and it works; it's visible in the app when opening the order. I also tried creating the shipping label from the app. In this specific case, the refresh does not work as you mentioned, but it also doesn't refresh in the `trunk` branch. I've asked in C02KUCFCSFP/p1738239651479579 if this is part of a new flow, since I remember the UI for creating the shipping label was different a while back. In any case, it's not something related to this PR. 👍

selanthiraiyan commented on Networking/Networking/Model/Order.swift:
```diff
@@ -239,6 +245,9 @@ public struct Order: Decodable, Sendable, GeneratedCopiable, GeneratedFakeable {
         }()
 
+        // Shipping labels
+        let shippingLabels = try container.decodeIfPresent([ShippingLabel].self, forKey: .shippingLabels) ?? []
```
👋 @pmusolino  Do we receive this `shipping_labels` key in orders response? Based on my testing in a Woo store with shipping labels, I don't see it being sent in `wc/v3/orders`. 

If we do receive the `shipping_labels`, should we update the mock JSON and add unit tests to ensure that `shipping_labels` are mapped correctly?

pmusolino replied:
Good question, @selanthiraiyan. In `ShippingLabelStore`, the method `upsertShippingLabels` connects the fetched shipping labels to the specific order, so the order doesn't return specific shipping labels, making it unnecessary to decode the Shipping Labels in the Order model. However, declaring the property (without decoding) in the networking model seems necessary to access `shippingLabels`, unless you have another solution in mind. 

Yes, we could access the shipping labels from storage, but that's what we're trying to avoid. Do you think I can leave the `shippingLabels` property in `Order.swift` but remove the decoding? This way, by default, it would be an empty array.

selanthiraiyan replied:
Thanks, @pmusolino! I don't have a solution in mind. But I think we should remove the decoding, as it will be misleading. 

I'm tagging @itsmeichigo to check whether she has any other solutions.

itsmeichigo replied:
Thanks for the ping Sharma!

I think if the key for shipping labels is non-existent in the order response, keeping the property an empty array should be the way to go. There's no need to waste more resources to look for such a key to decode. Please also leave a comment in the code to explain why the value is empty by default, and when it's going to be updated.

However, I'm concerned about this part:

> In ShippingLabelStore, the method upsertShippingLabels connects the fetched shipping labels to the specific order

In the order details screen, after syncing shipping labels, we'd need to ensure that the `order` in the view model is updated with the fetched label(s), just so they can be displayed on the UI.

pmusolino replied:
Thank you both! I updated the decoding here 85c0115c456026ec9e6865542d2513a3887c5d36 the comment should be clear enough.

> In the order details screen, after syncing shipping labels, we'd need to ensure that the order in the view model is updated with the fetched label(s), just so they can be displayed on the UI.

Yes, this already happens from my testing. If you create or update a shipping label, or open an order detail with an existing shipping label that needs to be fetched remotely, you will see the order detail UI correctly updated.

pmusolino committed:
Merge branch 'trunk' into feat/batch-fetching-in-order-details

pmusolino committed:
Merge branch 'trunk' into feat/batch-fetching-in-order-details

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/237">21.7</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

pmusolino committed:
Update Order.swift and RELEASE-NOTES.txt
- Modify `Order.swift` to initialize `shippingLabels` as an empty array by default.
- Update RELEASE-NOTES.txt to correctly place and describe the improved data fetching change in version 21.7.

selanthiraiyan reviewed (commented):
@pmusolino I cannot see the purchased shipping label in the order detail screen. (Even after multiple pull to refresh attempts)

https://github.com/user-attachments/assets/5b577455-fdc1-4b90-99df-3182adc70bd3

---

I tested by changing the Xcode scheme Build configuration into "Release" to ensure that we test the shipping labels work in production. (And not the WIP shipping labels project which is behind a feature flag)

<img width="1009" alt="image" src="https://github.com/user-attachments/assets/f49f4b2f-fddf-482d-9430-68a7a9fe7fd1" />

pmusolino commented:
@selanthiraiyan I was blocked from testing that flow because of this new issue https://github.com/woocommerce/woocommerce-ios/issues/15078 but I believe you didn't encounter this issue.

pmusolino commented:
@selanthiraiyan it seems to be an issue in `trunk` as well, and it also depends on which button you press at the end of the flow. If you press `Print Shipping Label`, the shipping label will appear in the order detail and on the web. However, if you press `Save for later`, I can't see or print the shipping label on mobile or on the web. The only thing visible on the web is the shipment tracking. In my tests, I always pressed "Print Shipping Label", but "Save for Later" seems to have a different behavior. I've pinged Eagle to understand if this is normal or an issue. C02KUCFCSFP/p1738850660636149

<img src="https://github.com/user-attachments/assets/22ea3457-acef-4058-a0f3-b2be51a3b4da" width=300 />

itsmeichigo commented:
👋 @pmusolino - jumping in because I have a question regarding this line:

> Those were the only relationships with Order.

I also found `refunds` as a relationship with order, maybe it should be cleaned up in this PR as well?

pmusolino commented:
@itsmeichigo Order Refund has not been cleaned up, because the `refunds` relationship in `Order` detail is of type `OrderRefundCondensed` while the expected one in `OrderDetailsResultsControllers` is of type `Refund`, which includes additional fields like `orderID`, `dateCreated`, `refundedByUserID`, `isAutomated`, etc… `OrderRefundCondensed` is suitable for scenarios where only basic refund info is needed. So, we still need to fetch the refunds from Core Data.

rachelmcr commented:
@pmusolino This isn't working well for me when I test with opening the order details for an order with a shipping label purchased on the web. Here's the scenario:

* I've opened the order in the app before (so it exists in local storage).
* I purchased a shipping label on the web.
* I open the order in the app again and expect the shipping label to be there.

In `trunk` this works as expected:

https://github.com/user-attachments/assets/908e4102-9b04-4a45-8c47-39640488d0d0

But in this branch, the shipping label shows up and then disappears. It only reappears after I pull to refresh:

https://github.com/user-attachments/assets/a6cb6bb6-1499-41b7-91ff-4859fdcd8a81

rachelmcr commented:
> If you press Print Shipping Label, the shipping label will appear in the order detail and on the web. However, if you press Save for later, I can't see or print the shipping label on mobile

FWIW I can't reproduce this behavior on my test store. Even if I select "Save for later," I see the purchased shipping label in order details in the app when I test in `trunk`:

https://github.com/user-attachments/assets/83f0d647-a7bb-4c9a-880f-0879a2f81244

I tested the same steps in this branch and it also worked for me there.

pmusolino commented:
Thank you @rachelmcr for testing it.

I discovered that there is a bug (also in `trunk`) where the shipping labels are fetched, but not displayed. I fixed the issue here https://github.com/woocommerce/woocommerce-ios/pull/15101 so I'm waiting it to be merged in `trunk` before re-testing this PR with what you found. 🙌

pmusolino committed:
Merge branch 'trunk' into feat/batch-fetching-in-order-details

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/238">21.8</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

pmusolino commented:
@rachelmcr I re-tested this PR after merging `trunk` back here, and I'm able to match the same behavior between `trunk` and this.

1) I opened an order without shipping labels.
2) Pressed back to return to the order list.
3) Created a shipping label on the web.
4) Re-opened the order detail: I'm able to see the shipping label created on the web.

Would you be able to take a second look, please?

rachelmcr commented:
@pmusolino I'm still seeing the same behavior as before. I haven't confirmed this theory, but I suspect it depends on whether we update the order details data source before the shipping labels are done being synced:

* In the view model, we [sync everything](https://github.com/woocommerce/woocommerce-ios/blob/a140b84a852ea51a7c8607219e786a876c79e654/WooCommerce/Classes/ViewModels/Order%20Details/OrderDetailsViewModel.swift#L226) (including the order and shipping labels).
* When we sync the order, if the order already exists in storage, we [use the last modified date to check if it has been updated remotely](https://github.com/woocommerce/woocommerce-ios/blob/a140b84a852ea51a7c8607219e786a876c79e654/Yosemite/Yosemite/Stores/OrderStore.swift#L248). (This date doesn't change if shipping labels are added to the order, since those are stored separately in the database and not considered part of the order. Related discussion: p7bje6-56E-p2 )
* We [only fetch the order from remote](https://github.com/woocommerce/woocommerce-ios/blob/a140b84a852ea51a7c8607219e786a876c79e654/Yosemite/Yosemite/Stores/OrderStore.swift#L253) if it has been updated since we last stored it locally.
* Once we finish syncing the order (which can be very fast if there are no remote changes) we immediately [update the order in order details](https://github.com/woocommerce/woocommerce-ios/blob/a140b84a852ea51a7c8607219e786a876c79e654/WooCommerce/Classes/ViewModels/Order%20Details/OrderDetailsViewModel.swift#L336).

If the remote sync for shipping labels hasn't finished yet at that point, the order details won't reflect the new shipping label(s) from remote. We probably need to ensure the last step to update the order in order details doesn't happen until after the shipping label sync is complete, or is updated again after the shipping labels finish syncing.

pmusolino committed:
Merge commit '7627219dad37276b64378daca7a1258fe46f90be' into feat/batch-fetching-in-order-details

pmusolino committed:
fix: correctly update shipping labels in an order if the following sequence of events occurs:

- I open an order detail without shipping labels on mobile and then navigate back to the list.
- I open the same order on the web, purchase a shipping label, and press the "update" button to refresh the order dates.
- When I reopen the order on mobile, the purchased shipping label appears while the order is updating. However, once the response is received, the shipping label disappears.

pmusolino commented:
@rachelmcr I apologize for my delayed response Rachel. I've been focusing on uploading product images in the background lately. Thank you for the explanation and the extra context! I was able to replicate the behavior, and the secret to replicate its not only to purchase the shipping label, but also returning to the order detail on the web and press "Update" button. I think I found a solution, and I created a branch [here](https://github.com/woocommerce/woocommerce-ios/tree/feat/batch-fetching-in-order-details-take-2) called `feat/batch-fetching-in-order-details-take-2`. I would appreciate it if you could test the changes and check the code to ensure that everything works correctly now. Tests are missing, but I will integrate and update the existing ones after you confirm that this solution works, merging the branch in this PR later. 🙌

From my discovery, the core problem was a race condition between order syncing and shipping label syncing.

**The issue was occurring because:**

1. When opening an order that exists in local storage, we were checking if the order was modified remotely.
2. If the order itself wasn't modified (as you said, adding shipping labels doesn't update the order's modification date), we would use the cached order.
3. The shipping label sync would happen in parallel, but the UI would already be updated with the cached order data.
4. When shipping labels finished syncing, there was no mechanism to update the order with those labels.

**So I applied the following changes:**
- In `ShippingLabelStore`, I modified `synchronizeShippingLabels` method to return the fetched shipping labels in its completion handler instead of just a void success result.
- In `ShippingLabelAction`, I updated the `synchronizeShippingLabels` action definition to match the store implementation and pass the shipping labels in the completion handler.
- In `OrderDetailsViewModel`, I updated `syncShippingLabels()` to return the shipping labels when available and I also restructured the sync process in `syncEverything()` by moving the shipping label sync task inside the order sync group, then I added explicit code to update the order with the newly synced shipping labels and in the end a UI reload.


Let me know if it works for you 🙇‍♂️

selanthiraiyan commented:
@pmusolino I tested the code from the branch [feat/batch-fetching-in-order-details-take-2](https://github.com/woocommerce/woocommerce-ios/tree/feat/batch-fetching-in-order-details-take-2), and I can see the shipping label purchased from the web without using pull to refresh in the order details screen.

1. I noticed that the "Create Shipping Label" is visible momentarily before the purchased shipping label is displayed. 
2. In terms of code, I am concerned about us manually updating the order with the synched shipping labels. [Code](https://github.com/woocommerce/woocommerce-ios/blob/35773814a9d8f129f9ec763661546d306e724e2d/WooCommerce/Classes/ViewModels/Order%20Details/OrderDetailsViewModel.swift#L282-L285) I think we should listen to the shipping labels from the storage and update the order. Is that possible?

pmusolino commented:
Thanks, Sharma, for checking the branch!

> 1. I noticed that the "Create Shipping Label" is visible momentarily before the purchased shipping label is displayed. 

I believe this is one of the issues with the order details. During loading, there are several glitches because the views update after multiple fetches. I will check if this problem was also present earlier. 👍

> 2. In terms of code, I am concerned about us manually updating the order with the synched shipping labels. [Code](https://github.com/woocommerce/woocommerce-ios/blob/35773814a9d8f129f9ec763661546d306e724e2d/WooCommerce/Classes/ViewModels/Order%20Details/OrderDetailsViewModel.swift#L282-L285) I think we should listen to the shipping labels from the storage and update the order. Is that possible?

In this PR, we are trying to avoid reading from storage, as it contributes to the `performFetch()` issue. Therefore, monitoring shipping labels from storage seems far from the ideal approach, otherwise, it's like shifting the problem from explicit fetching to monitoring. Wdyt?

selanthiraiyan commented:
> In this PR, we are trying to avoid reading from storage, as it contributes to the performFetch() issue. Therefore, monitoring shipping labels from storage seems far from the ideal approach, otherwise, it's like shifting the problem from explicit fetching to monitoring. Wdyt?

👋 @pmusolino,
As you replied [here](https://github.com/woocommerce/woocommerce-ios/pull/14999#discussion_r1943227552) in the discussion with @itsmeichigo, will the `order` in the view model automatically be updated once the shipping labels are synched? If yes, please help me understand why we have to manually update the order with the synched labels. Thanks for your patience!

pmusolino committed:
Merge commit 'd231feff7094c39c56123f4fd2a065bcf459ca6f' into feat/batch-fetching-in-order-details-take-2

pmusolino committed:
Merge commit 'd231feff7094c39c56123f4fd2a065bcf459ca6f' into feat/batch-fetching-in-order-details

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/239">21.9</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

pmusolino commented:
@selanthiraiyan when I replied in that discussion with Huong, I wasn't aware of the problem mentioned [here](https://github.com/woocommerce/woocommerce-ios/pull/14999#issuecomment-2671911263) by Rachel.

Why?
In my testing, I was always pressing "Update" for updating the order after creating a shipping label, while I discovered this is not needed for generating and attaching a shipping label on the web (In fact, I believe it is a web issue because the Update button is enabled even if there are no changes). 
Pressing "Update", changes the behavior of the refresh of the shipping labels, and the reason is explained in this Rachel's comment: 

> _When we sync the order, if the order already exists in storage, we [use the last modified date to check if it has been updated remotely](https://github.com/woocommerce/woocommerce-ios/blob/a140b84a852ea51a7c8607219e786a876c79e654/Yosemite/Yosemite/Stores/OrderStore.swift#L248). (This date doesn't change if shipping labels are added to the order, since those are stored separately in the database and not considered part of the order_

So, when you create a shipping label from the web without saving the order, the order's modification date does not update. As a result, everything does not refresh correctly because the shipping label and the order are treated as two separate entities. However, if you refresh the order, the shipping labels will be attached correctly.

When we get the shipping labels from the syncShippingLabels() method, now we need to explicitly attach them to the order using:
```
if let updatedOrder = self?.order.copy(shippingLabels: shippingLabels) {
    self?.update(order: updatedOrder)
}
```
Previously, this was unnecessary because we were retrieving data from storage.
Without this, the UI would continue to use the existing order object in memory, which is not updated.

Let me know if I have cleared your doubts, Sharma. 🙌

selanthiraiyan commented:
Thanks for the detailed answer, @pmusolino!

Would it work to sync the shipping labels first before loading the order from the storage?

1. Send the remote request to fetch shipping labels.
2. Update the order in storage with the shipping label.
3. Send a request to retrieve the order. 
4. If the order [has not been modified recently](https://github.com/woocommerce/woocommerce-ios/blob/a140b84a852ea51a7c8607219e786a876c79e654/Yosemite/Yosemite/Stores/OrderStore.swift#L248), we will still end up loading the latest order with shipping labels from storage.

In the above approach,
- We also avoid manually updating the `Order` in the view layer. 
- I assume we can also avoid showing the "Create Shipping Labels" momentarily.

pmusolino commented:
@selanthiraiyan Not sure this is the ideal solution because the core of Order Detail is the sync of the order in `OrderDetailsViewModel`. From there, there is a list of nested operations, including shipping labels. So, the initial info of the order is displayed, and other sections are loaded later. If we first load Shipping Labels, we create a series of dependencies in reverse because we should first check the plugins to see if the shipping labels are supported, then fetch the shipping labels if they are supported, and then fetch the order to retrieve it for the first time or update it. This would also complicate the logic, seemingly with no advantages, and, in fact, it might worsen the experience at the UI level (eg. now there is a bar displaying that the order is still loading, but the main info is already displayed).

selanthiraiyan commented:
Thanks for the explanation, @pmusolino! I understand. Manually updating the order with the synched shipping labels seems like the way to go. Do you plan to add those changes to this PR?

pmusolino commented:
@selanthiraiyan if you think the work I did here https://github.com/woocommerce/woocommerce-ios/compare/feat/batch-fetching-in-order-details...feat/batch-fetching-in-order-details-take-2 works, I can merge the changes in this PR 🙌

selanthiraiyan commented:
>  I can merge the changes in this PR 🙌

@pmusolino Yes, please go ahead with this. Thanks!

pmusolino committed:
Merge branch 'feat/batch-fetching-in-order-details-take-2' into feat/batch-fetching-in-order-details

pmusolino committed:
Merge branch 'trunk' into feat/batch-fetching-in-order-details

pmusolino committed:
update: release-notes

pmusolino commented:
@selanthiraiyan done ✅ waiting for the final approval

selanthiraiyan commented on WooCommerce/Classes/ViewModels/Order Details/OrderDetailsViewModel.swift:
```diff
@@ -262,6 +262,32 @@ extension OrderDetailsViewModel {
+                    group.leave()
+                }
+                // Check Woo Shipping support first, to ensure correct flows are enabled for shipping labels.
+                self?.dataSource.isEligibleForWooShipping = ((await self?.isWooShippingSupported()) != nil)
```
We could unwrap `self` using a guard statement to avoid using self as an optional here and a few other places below.

itsmeichigo replied:
Done in bbfb18c0b00fd7ba7a54318eb920f8ec66e13d92.

selanthiraiyan commented on WooCommerce/Classes/ViewModels/Order Details/OrderDetailsViewModel.swift:
```diff
@@ -681,24 +691,24 @@ extension OrderDetailsViewModel {
 
     @MainActor
-    func syncShippingLabels() async {
+    func syncShippingLabels() async -> [ShippingLabel]? {
```
This method can be marked as `private` as we don't access this from outside of this view model.

selanthiraiyan replied:
Marking as resolved as we access this method from unit tests.

selanthiraiyan commented on WooCommerce/Classes/ViewModels/Order Details/OrderDetailsViewModel.swift:
```diff
@@ -262,6 +262,32 @@ extension OrderDetailsViewModel {
+                self?.dataSource.isEligibleForShippingLabelCreation = isEligible ?? false
+
+                // Sync shipping labels and update order with the result if available
+                if let shippingLabels = await self?.syncShippingLabels() {
```
Should we call `syncShippingLabels` even if `checkShippingLabelCreationEligibility` returned false above? If yes, can we start this request asynchronously without waiting for the response from `checkShippingLabelCreationEligibility`?

itsmeichigo replied:
We still need to sync shipping labels just in case the creation eligibility is only updated recently. I moved these calls to a task group to ensure they don't block each other in bbfb18c0b00fd7ba7a54318eb920f8ec66e13d92.

selanthiraiyan reviewed (commented):
Thanks for your patience with this work, Paolo! ❤️ 


#### Code

I left a few minor comments and one question about the code. We are getting pretty close to merging this PR. 👏 

#### Testing

1. I validated that a label purchased from web is shown in the order details page. [This test case scenario.](https://github.com/woocommerce/woocommerce-ios/pull/14999#issuecomment-2642757284) ✅ 
2. The order details page doesn't reflect a newly purchased shipping label unless I pull down to refresh. [Mentioned in this previous comment](https://github.com/woocommerce/woocommerce-ios/pull/14999#pullrequestreview-2583040307) I tested that this bug exists in trunk as well. ✅ 

---

👋  @itsmeichigo It would be great if you could you take a look at this PR as a confidence check? Thanks! 🙇

itsmeichigo committed:
Fix issue triggering order syncing after dismissing modals

itsmeichigo commented:
Thanks for the ping @selanthiraiyan! I don't have any additional comment aside from your latest ones. 

Regarding this point:

> The order details page doesn't reflect a newly purchased shipping label unless I pull down to refresh. https://github.com/woocommerce/woocommerce-ios/pull/14999#pullrequestreview-2583040307 I tested that this bug exists in trunk as well.

I did some debugging, and this seems to be an issue with `viewWillAppear` not being triggered after dismissing the shipping label creation flow. This was caused by the shipping label creation flow being presented with `overFullScreen` mode. 

I fixed this by switching to `fullScreen` mode for both the shipping label creation flow and order edit flow in 32dbb12df1fda1f03ed381eb4c0753e3ca78c0e4. This ensures that the order is synced again after dismissing these flows. The created shipping labels should now be displayed after dismissing the creation flow.

itsmeichigo committed:
Fix unit test build failures

itsmeichigo committed:
Improve syncing of shipping label details

itsmeichigo commented:
@selanthiraiyan I added the final changes following your latest comments, please take another look when you can.

selanthiraiyan reviewed (approved):
Thanks for handling this, Huong! The changes LGTM. 

Let's merge this PR. 🚀

pmusolino commented:
Thank you, Huong, for taking care of the remaining changes! 🙏
